### PR TITLE
bugfix: close keyboard on scroll

### PR DESCRIPTION
**Pull request by simplenotezy**
A minor update, that hides the keyboard when scrolling the list:

https://user-images.githubusercontent.com/3786627/220783239-c1db1f96-7608-4487-8942-193d1b240bb8.MP4

### CODE REVIEW

**Comment by cedvdb** [comment]
This is a fantastic UX improvement. Thanks

File: lib/src/widgets/country_selector/country_selector_navigator.dart [hunk 1]
```diff
@@ -428,6 +428,11 @@ class DraggableModalBottomSheetNavigator extends CountrySelectorNavigator {
         maxChildSize: maxChildSize,
         expand: false,
         builder: (context, scrollController) {
+          // hide the keyboard when scrolling the list, so it doesn't overlap the list
+          scrollController.addListener(() {
```
Comment (cedvdb) [inline lib/src/widgets/country_selector/country_selector_navigator.dart:432]: Is the scroll controller disposed somewhere, it does not seem to be in this PR ? 

  1. It's a bit odd to have the listener added here and removed somewhere else
  2. I believe this will add a listener on every rebuild

Comment (cedvdb) [reply lib/src/widgets/country_selector/country_selector_navigator.dart:432]: @simplenotezy do you plan on resolving this in the short term ? Asking for a future release

Comment (simplenotezy) [reply lib/src/widgets/country_selector/country_selector_navigator.dart]: @cedvdb excuse me for the latency 😅 however, now we're removing the listener on dispose. See latest commit. This was the best way I could think of for now (creating a new widget, so we can get access to the lifecycle methods)

Comment (simplenotezy) [reply lib/src/widgets/country_selector/country_selector_navigator.dart]: I have also aligned the PR with dev branch so I think it's safe to resolve this conversation now

**Comment by cedvdb** [comment]
Same as for the other PR Let's add a test and remove the part that's for another PR please. This is a great addition thanks again.

File: lib/src/widgets/phone_field_state.dart [hunk 2]
```diff
@@ -40,6 +40,15 @@ class PhoneFieldState extends State<PhoneField> {
     SystemChannels.textInput.invokeMethod('TextInput.show');
   }
 
+  // TODO: Would be cleaner if we could infer it from
```
[the remark this answers is not in the document]
Comment (simplenotezy) [reply lib/src/widgets/phone_field_state.dart:43]: Yup, you're right. My mistake merging it in here.

**Comment by simplenotezy** [comment]
@cedvdb I don't have any experience with writing such tests in Flutter unfortunately, and very limited on time these days, would learn a lot by seeing someone else present how such a frontend test could look like.

File: lib/src/widgets/country_selector/country_selector_navigator.dart (hunk 1 above)
Comment (simplenotezy) [reply lib/src/widgets/country_selector/country_selector_navigator.dart]: @cedvdb by the way, listeners are automatically cleared on dispose:

image

**Comment by cedvdb** [comment]
> @cedvdb I don't have any experience with writing such tests in Flutter unfortunately, and very limited on time these days, would learn a lot by seeing someone else present how such a frontend test could look like.

There are tests in the test folder.  When scrolling the keyboard is not shown anymore would be good.

Essentially something like the pseudo code
 
```dart
testWidgets('Should hide keyboard on scroll', (widgetTester) {

    await widgetTester.pumpWidget( country-selector-widget );
    await widgetTester.pumpAndSettle();

    // test that the keyboard is shown initially
    expect(tester.testTextInput.isVisible, isTrue);

    await widgetTester.scroll(...);

    await widgetTester.pumpAndSettle();
    // check that the keyboard is not shown anymore
    expect(tester.testTextInput.isVisible, isFalse);

});
```

File: lib/src/widgets/country_selector/search_box.dart [hunk 3]
```diff
@@ -18,21 +18,49 @@ class SearchBox extends StatelessWidget {
...
+  void initState() {
+    super.initState();
+  }
+
+  void handleChange(e) {
+    widget.onChanged(e);
+
+    // detect length difference
+    final diff = e.length - _previousValue.length;
+    if (diff > 3) {
```
Comment (cedvdb) [inline lib/src/widgets/country_selector/search_box.dart:38]: note that this change seems to be for another PR.

**Comment by cedvdb** [comment]
@simplenotezy Is the comment above enough to go forward with the test ?
